pkg/operator/controller/route-metrics/controller.go

```diff
@@ -0,0 +1,191 @@
+		return nil, err
+	}
+	// add watch for changes in Route
+	if err := c.Watch(source.NewKindWithCache(&routev1.Route{}, newCache),
+		handler.EnqueueRequestsFromMapFunc(reconciler.routeToIngressController)); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// routeToIngressController creates a reconcile.Request for all the Ingress Controllers related to the Route object.
+func (r *reconciler) routeToIngressController(obj client.Object) []reconcile.Request {
+	var requests []reconcile.Request
+	// Cast the received object into Route object.
+	routeObject := obj.(*routev1.Route)
+
+	// Iterate through the related RouteIngresses.
+	for _, ri := range routeObject.Status.Ingress {
+		// Check if the Route was admitted by the RouteIngress.
+		for _, cond := range ri.Conditions {
+			if cond.Type == routev1.RouteAdmitted && cond.Status == corev1.ConditionTrue {
+				log.Info("queueing ingresscontroller", "name", ri.RouterName)
+				// Create a reconcile.Request for the router named in the RouteIngress.
+				request := reconcile.Request{
+					NamespacedName: types.NamespacedName{
+						Name:      ri.RouterName,
+						Namespace: r.Namespace,
+					},
+				}
+				requests = append(requests, request)
+			}
+		}
+	}
+	return requests
+}
+
+// reconciler handles the actual ingresscontroller reconciliation logic in response to events.
+type reconciler struct {
+	cache     cache.Cache
+	Namespace string
+}
+
+// Reconcile expects request to refer to a Ingress Controller resource, and will do all the work to gather metrics related to
+// the resource.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log.Info("reconciling", "request", request)
+
+	// Fetch the Ingress Controller object.
+	ingressController := &operatorv1.IngressController{}
+	if err := r.cache.Get(ctx, request.NamespacedName, ingressController); err != nil {
+		if kerrors.IsNotFound(err) {
+			// This means the Ingress Controller object was already deleted/finalized.
+			log.Info("Ingress Controller not found; reconciliation will be skipped", "request", request)
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("failed to get Ingress Controller %q: %v", request, err)
+	}
+
+	// If the ingresscontroller is deleted, then delete the corresponding RouteMetricsControllerRoutesPerShard metric label
+	// and return early.
+	if ingressController.DeletionTimestamp != nil {
+		// Delete the Shard label corresponding to the Ingress Controller from the RouteMetricsControllerRoutesPerShard metric.
+		DeleteRouteMetricsControllerRoutesPerShardMetric(request.Name)
+		log.Info("RoutesPerShard metric label corresponding to the Ingress Controller is successfully deleted", "ingresscontroller", ingressController)
+		return reconcile.Result{}, nil
+	}
+
+	// Get the Namespace LabelSelector from the Ingress Controller.
```

Comment:
Generally, the route admitted status should take into account the namespace label selectors (it wouldn't have been admitted if it wasn't in the right namespace).

Now, I do see 1 scenario why you would want to check namespace labels: if the namespace object has it's label changed and was no longer admitted by Ingress Controller because of this, the Route Status will most likely still show admitted as this is a known bug after doing https://bugzilla.redhat.com/show_bug.cgi?id=1944851. 

So, after all, I think checking the namespace selector may be a good idea, but I just want to make sure we are on the same page, and we should probably write a comment about this along the lines:

`\\ Even though the route admitted status should reflect validity of the namespace labelselector, we still will validate the namespace as there are still edge scenarios where the route status may be inaccurate`

Open to discussion on this one too.

Reply:
I actually see a different problem with the Route Status being cleared. Whenever a Route event occurs, the reconcile runs only for the current Route Ingresses which has admitted Route. If the event was raised due to Route Status being cleared for a Route Ingress due to update of the label of the Route or the Namespace it was created in, then the metric label for the Route Ingress which got removed from the Status will not be updated. We have to reconcile for all Ingress Controllers whenever a Route event occurs.

Reply:
> Now, I do see 1 scenario why you would want to check namespace labels: if the namespace object has it's label changed and was no longer admitted by Ingress Controller because of this, the Route Status will most likely still show admitted as this is a known bug after doing https://bugzilla.redhat.com/show_bug.cgi?id=1944851.

Is it only an issue with the status not be cleared or will the haproxy.config still have an entry for the Route as well? If it's only a problem of the status not being cleared then it would be fine to use the namespace labelselector as the Route will not have any effect on the traffic load of the Router. Then adding the comment makes sense.

Reply:
> Whenever a Route event occurs, the reconcile runs only for the current Route Ingresses which has admitted Route.

That's a good point. This logic doesn't capture "opting-out" of a route shard via a route or namespace label. I think the proper way to do this is `routeToIngressController` be removed and we reconcile all ingress controllers on any route or ingress controller update. That seems much more computation intensive, but I don't know if I have any better ideas. Thoughts?

> Is it only an issue with the status not be cleared or will the haproxy.config still have an entry for the Route as well?

Just confirm that it haproxy.config will not have a entry for the route after a namespace label match removal, but the route status will still be there. So the status isn't completely in sync, but I just wanted to make that clear. Some sort of comment addressing this may be helpful for the future like I mentioned. I think this should be written up as a bug at some point.

Reply:
I wrote the bug up https://issues.redhat.com/browse/OCPBUGS-1689

Reply:
> That's a good point. This logic doesn't capture "opting-out" of a route shard via a route or namespace label. I think the proper way to do this is `routeToIngressController` be removed and we reconcile all ingress controllers on any route or ingress controller update. That seems much more computation intensive, but I don't know if I have any better ideas. Thoughts?

We can keep track of the Route Ingresses for each Route by using a local variable (it can be a map of Route Name to Set of Route Ingress Names). This variable will be updated every reconcile and it'll help us to find out if a Route Ingress was added or removed. Based on that only the Ingress Controller that `changed` can be reconciled.


> Some sort of comment addressing this may be helpful for the future like I mentioned.

Yes. The comment will give a better context.

Reply:
I've pushed the change. PLMK if it's fine. If not, I will rollback the change.

Reply:
If we still have a bug here, it should be fixed before this goes to QA.